pyramid_tm/__init__.py

```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+
+                # check for a squashed exception and handle it
+                # this would happen if an exception view was invoked and
+                # rendered an error response
+                exc_info = getattr(request, 'exc_info', None)
+                if exc_info is not None:
+                    # if this was the last attempt or the exception is not
+                    # retryable use this response instead of raising and
+                    # having a new error response generated
+                    if (
+                        is_last_attempt or
+                        not is_exc_retryable(request, exc_info)
+                    ):
+                        raise AbortWithResponse(response)
+
+                    # the exception is retryable so we'll squash the response
+                    # and loop around to try again
+                    else:
+                        reraise(*exc_info)
+
                 if commit_veto is not None:
                     veto = commit_veto(request, response)
                     if veto:
-                        raise AbortResponse(response)
+                        raise AbortWithResponse(response)
                 manager.commit()
-                del request.environ['tm.active']
                 return response
-            except AbortResponse as e:
+
+            except AbortWithResponse as e:
                 manager.abort()
-                del request.environ['tm.active']
                 return e.response
-            except:
+
+            except Exception:
                 exc_info = sys.exc_info()
                 try:
-                    retryable = manager._retryable(*exc_info[:-1])
-                    manager.abort()
-                    if (number <= 0) or (not retryable):
-                        del request.environ['tm.active']
-                        reraise(*exc_info)
+                    # if this was the last attempt or the exception is not
+                    # retryable then make a last ditch effort to render an
+                    # error response before sending the exception up the stack
+                    if (
+                        is_last_attempt or
+                        not is_exc_retryable(request, exc_info)
+                    ):
+                        return render_exception(request, exc_info)
+
                 finally:
+                    # keep the manager alive until after invoke_exception_view
+                    manager.abort()
+
                     del exc_info # avoid leak
 
+            # cleanup any changes we made to the request
+            finally:
+                request_manager.pop()
+
+                del environ['tm.active']
+                del environ['tm.attempt']
+                del environ['tm.attempts']
+
+                # propagate exception info back to the original request,
+                # possibly clearing out a retryable error triggered from the
+                # first attempt
```

Comment:
:-( I understand why, but this seems like it can be error prone... how about any finished callbacks and all that stuff? Those are set on this latest request, not on the original request...

You are right when you said the callbacks need to be under pyramid_tm.